core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/KOptDescriptor.java

```diff
@@ -0,0 +1,271 @@
+package org.optaplanner.core.impl.heuristic.selector.move.generic.list;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.optaplanner.core.api.function.TriPredicate;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.index.IndexVariableSupply;
+import org.optaplanner.core.impl.util.Pair;
+
+public final class KOptDescriptor<Solution_> {
+
+    /**
+     * The number of edges being added
+     */
+    private final int k;
+
+    /**
+     * A sequence of 2K nodes (t_0, t_1, ..., t_2K) where:
+     * (t_(2i - 1), t_(2i)) is an out-edge (an edge to be deleted)
+     * (t_(2i), t_(2i+1)) is an in-edge (an edge to be added)
+     */
+    private final Object[] t;
+
+    /**
+     * The order each node is visited when the tour is travelled in the successor direction
+     */
+    private final Integer[] p;
+
+    /**
+     * The order each node is visited when the tour is travelled in the successor direction. It the
+     * inverse of {@link KOptDescriptor#p} (i.e.
+     * {@link KOptDescriptor#p}[predecessorVisitingOrder[i]] == i
+     *
+     */
+    private final Integer[] q;
+
+    private final Integer[] incl;
```

Comment:
Any particular reason why these are not `int[]`? Do you expect `null` values in either?
From the code, it doesn't appear so.

Reply:
Cannot sort using a custom Comparator with a primitive array. Sorting is done to determine the order each element would be encountered if travelled in the successor direction (starting from t[1]).